geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloFeatureReader.scala

```diff
@@ -104,30 +102,27 @@ class AccumuloFeatureReaderWithStats(val query: Query,
   }
 }
 
-private class AccumuloFeatureReaderDelegate(iter: SFIter, query: Query, qp: QueryPlanner)(implicit timings: Timings)
-    extends FeatureReader[SimpleFeatureType, SimpleFeature] with MethodProfiling {
+trait FeatureCounting extends AccumuloFeatureReader {
 
-  override def getFeatureType: SimpleFeatureType = query.getHints.getReturnSft
-  override def next(): SimpleFeature = profile(iter.next(), "next")
-  override def hasNext: Boolean = profile(iter.hasNext, "hasNext")
-  override def close(): Unit = iter.close()
-
-  def getCount: Int = timings.occurrences("next").toInt
-}
-
-private class AccumuloFeatureReaderBinDelegate(iter: SFIter, query: Query, qp: QueryPlanner)(implicit timings: Timings)
-    extends AccumuloFeatureReaderDelegate(iter, query, qp)(timings) {
-
-  import org.locationtech.geomesa.accumulo.iterators.BinAggregatingIterator.BIN_ATTRIBUTE_INDEX
-
-  private var count = 0
-  private val bytesPerHit = if (query.getHints.getBinLabelField.isDefined) 24 else 16
+  protected var counter = 0L
+  // because the query planner configures the query hints, we can't check for bin hints
+  // until after setting up the iterator
+  protected val sfCount: (SimpleFeature) => Int = if (query.getHints.isBinQuery) {
+    val bytesPerHit = if (query.getHints.getBinLabelField.isDefined) 24 else 16
+    (sf) => sf.getAttribute(BIN_ATTRIBUTE_INDEX).asInstanceOf[Array[Byte]].length / bytesPerHit
+  } else {
+    (_) => 1
+  }
 
-  override def next(): SimpleFeature = {
+  abstract override def next(): SimpleFeature = {
     val sf = super.next()
-    count += (sf.getAttribute(BIN_ATTRIBUTE_INDEX).asInstanceOf[Array[Byte]].length / bytesPerHit)
+    counter += sfCount(sf)
     sf
   }
 
-  override def getCount: Int = count
+  abstract override def count = counter
+}
+
+trait FeatureLimiting extends FeatureCounting {
+  abstract override def hasNext: Boolean = counter < maxFeatures && super.hasNext
```

Comment:
Hmm, yeah. i think we'd have to check `hasNext` in each call to `next` to manage that... probably not worth it.